checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java

```diff
@@ -289,32 +291,58 @@ protected void checkConstructorResult(
-        for (AnnotationMirror classAnno : classAnnotations) {
-            for (AnnotationMirror constructorAnno : constructorAnnotations) {
-                if (qualifierHierarchy.leastUpperBound(classAnno, constructorAnno) != null) {
-                    if (qualifierHierarchy.isSubtype(constructorAnno, classAnno)) {
-                        if (!AnnotationUtils.areSameByName(constructorAnno, classAnno)) {
-                            checker.report(
-                                    Result.warning(
-                                            "inconsistent.constructor.type", constructorAnno),
-                                    constructorElement);
-                        }
-                    }
-                    break;
+        super.checkConstructorResult(constructorType, constructorElement);
+    }
+
+    @Override
+    public boolean validateTypeOf(Tree tree) {
+        // Don't check the result type of a constructor, because it must be @UnknownInterned, even
+        // if the type on the class declaration is @Interned.
+        if (tree.getKind() == Tree.Kind.METHOD && TreeUtils.isConstructor((MethodTree) tree)) {
+            return true;
+        } else if (tree.getKind() == Tree.Kind.NEW_CLASS) {
+            NewClassTree newClassTree = (NewClassTree) tree;
+            Element elt = TreeUtils.elementFromUse(newClassTree).getEnclosingElement();
+            Set<AnnotationMirror> bounds = atypeFactory.getTypeDeclarationBounds((TypeElement) elt);
+            // Don't issue an invalid type warning for creations of objects of interned classes;
+            // instead, issue an interned.object.creation if required.
+            if (AnnotationUtils.containsSameByClass(bounds, Interned.class)) {
+                ParameterizedExecutableType fromUse = atypeFactory.constructorFromUse(newClassTree);
+                AnnotatedExecutableType constructor = fromUse.executableType;
+                if (!checkCreationOfInternedObject(newClassTree, constructor)) {
+                    return false;
+                }
+            }
+        }
+        return super.validateTypeOf(tree);
+    }
+
+    /**
+     * Issue an error if {@code newInternedObject} is not immediately interned.
+     *
+     * @param newInternedObject call to a constructor of an interned class
+     * @param constructor declared type of the constructor
+     * @return false unless {@code newInternedObject} is immediately interned.
+     */
+    private boolean checkCreationOfInternedObject(
+            NewClassTree newInternedObject, AnnotatedExecutableType constructor) {
+        if (constructor.getReturnType().hasAnnotation(Interned.class)) {
+            return true;
+        }
+        if (getCurrentPath() != null
+                && getCurrentPath().getParentPath() != null
+                && getCurrentPath().getParentPath().getParentPath() != null) {
+            Tree parent = getCurrentPath().getParentPath().getParentPath().getLeaf();
```

Comment:
Can you add a comment why there are two parents to climb?
I assume this is to handle `myIntern(new MyInternType())`. Why is it not just one level to climb to get from the `new` tree to the `myIntern` tree?

Reply:
Oh, or is it for `new MyInternType().intern()`? So a comment would definitely be useful.

Reply:
I added a comment below.